KaitaiStream.cs

```diff
@@ -478,74 +488,150 @@ public static byte[] BytesTerminate(byte[] src, byte terminator, bool includeTer
+        /// Perform XOR processing between given data and single-byte key.
+        /// WARNING: May return same byte array if key is zero.
         /// </summary>
-        /// <param name="value">The data toe process</param>
-        /// <param name="key">The key value to XOR with</param>
-        /// <returns>Processed data</returns>
-        public byte[] ProcessXor(byte[] value, int key)
+        /// <param name="data">The data to process, as byte array</param>
+        /// <param name="key">The key to XOR with, as integer</param>
+        public byte[] ProcessXor(byte[] data, int key)
         {
-            byte[] result = new byte[value.Length];
-            for (int i = 0; i < value.Length; i++)
+            if (key == 0)
+                return data;
+
+            int dl = data.Length;
+            byte[] result = new byte[dl];
+
+            for (int i = 0; i < dl; i++)
             {
-                result[i] = (byte)(value[i] ^ key);
+                result[i] = (byte)(data[i] ^ key);
             }
             return result;
         }
 
         /// <summary>
-        /// Performs XOR processing with given data, XORing every byte of the input with a key
-        /// array, repeating from the beginning of the key array if necessary
+        /// Perform XOR processing between given data and multiple-byte key.
+        /// WARNING: May return same byte array if key is zero.
         /// </summary>
-        /// <param name="value">The data toe process</param>
-        /// <param name="key">The key array to XOR with</param>
-        /// <returns>Processed data</returns>
-        public byte[] ProcessXor(byte[] value, byte[] key)
+        /// <param name="data">The data to process, as byte array</param>
+        /// <param name="key">The key to XOR with, as byte array</param>
+        public byte[] ProcessXor(byte[] data, byte[] key)
         {
-            int keyLen = key.Length;
-            byte[] result = new byte[value.Length];
-            for (int i = 0, j = 0; i < value.Length; i++, j = (j + 1) % keyLen)
+            if (key.Length == 1)
+                return ProcessXor(data, key[0]);
+            if (key.Length <= 64 && ByteArrayZero(key))
+                return data;
+
+            int dl = data.Length;
+            int kl = key.Length;
+            byte[] result = new byte[dl];
+
+            for (int i = 0; i < dl; i++)
             {
-                result[i] = (byte)(value[i] ^ key[j]);
+                result[i] = (byte)(data[i] ^ key[i % kl]);
             }
             return result;
         }
 
         /// <summary>
-        /// Performs a circular left rotation shift for a given buffer by a given amount of bits.
+        /// Used internally.
+        /// </summary>
+        private static byte[][] precomputed_single_rotations;
+
+        /// <summary>
+        /// Used internally.
+        /// </summary>
+        private static void compute_single_rotations()
```

Comment:
`lower_underscore` is totally against any .NET coding standards, I'm afraid.

Reply:
Agree again. Best to not use underscores in C# code **precomputedSingleRotations** woud be better.

Reply:
Sorry, just copied the name from Python. It was a literal translation. Fixed.